File: src/lib/api/designContract.ts
import { invoke } from "@tauri-apps/api/core";

// ────────────────────────── Design Contract Types ──────────────────────────

export interface FontSize {
  name: string;
  size: string;
  lineHeight: string;
}

export interface DesignColors {
  primary: string;
  primaryHover: string;
  background: string;
  surface: string;
  textPrimary: string;
  textMuted: string;
  accent: string;
  success: string;
  warning: string;
  error: string;
  border: string;
  custom: Record<string, string>;
}

export interface DesignTypography {
  fontFamilyBase: string;
  fontFamilyHeading: string;
  fontFamilyMono: string;
  fontWeights: number[];
  sizeScale: FontSize[];
}

export interface DesignSpacing {
  baseUnit: number;
  scale: number[];
}

export interface ShadowLevel {
  name: string;
  value: string;
}

export interface DesignElevation {
  levels: ShadowLevel[];
}

export interface DesignShapes {
  borderRadius: Record<string, string>;
}

export interface ComponentSpec {
  name: string;
  description: string;
  variants: string[];
  sizes: string[];
  rules: string[];
}

export interface DesignGuardrail {
  rule: string;
  category: string;
  severity: string; // must | should | must_not | should_not
}

export interface DesignContract {
  schemaVersion: number;
  name: string;
  description?: string | null;
  colors: DesignColors;
  typography: DesignTypography;
  spacing: DesignSpacing;
  elevation: DesignElevation;
  shapes: DesignShapes;
  components: ComponentSpec[];
  guardrails: DesignGuardrail[];
  sourceTemplate?: string | null;
  sourceReference?: string | null;
  generatedAt: string;
  opensunstarVersion: string;
}

export interface DesignContractParams {
  templateId?: string | null;
  name: string;
  prototypeTemplate?: string | null;
  description?: string | null;
  colors?: DesignColors | null;
  typography?: DesignTypography | null;
  spacing?: DesignSpacing | null;
  elevation?: DesignElevation | null;
  shapes?: DesignShapes | null;
  components?: ComponentSpec[] | null;
  guardrails?: DesignGuardrail[] | null;
}

export interface ImportResult {
  contract: DesignContract;
  source: string;
  warnings: string[];
  quality: {
    level: "ready" | "needs_review";
    missingSections: string[];
  };
}

export interface DesignInstallResult {
  filesCreated: string[];
  filesSkipped: string[];
  designMdCreated: boolean;
  dtchgJsonCreated: boolean;
  manifestCreated: boolean;
}

export interface DesignSystemOutputVerification {
  path: string;
  expectedSha256: string;
  actualSha256?: string | null;
  state: "verified" | "missing" | "drifted";
}

export interface DesignSystemVerification {
  state: "verified" | "missing" | "invalid" | "drifted";
  outputs: DesignSystemOutputVerification[];
}

// ────────────────────────── Install Plan Types (shared) ──────────────────────────

export interface InstallFileEntry {
  path: string;
  status: "create" | "skip" | "overwrite";
  newContent?: string | null;
  existingContent?: string | null;
}

export interface InstallAuditFinding {
  severity: string;
  ruleId: string;
  message: string;
  file: string;
}

export interface InstallAuditSummary {
  filesScanned: number;
  totalFindings: number;
  critical: number;
  high: number;
  medium: number;
  low: number;
  blocked: boolean;
  findings: InstallAuditFinding[];
}

export interface DesignInstallPlan {
  files: InstallFileEntry[];
  audit: InstallAuditSummary;
}

export interface DesignSystemPackage {
  id: string; name: string; version: string; licenseId: string;
  applicableScenarios: string[]; source: "builtin" | "user";
}
export interface DesignSystemDiscovery { packages: DesignSystemPackage[]; rejected: { path: string; reason: string }[]; }
export interface DesignSystemPackageDetail { package: DesignSystemPackage; components: { components?: string[]; pageTemplates?: string[] }; responsive: { modes?: string[]; breakpoints?: Record<string, number>; rules?: string[] }; accessibility: string; }

// ────────────────────────── API Methods ──────────────────────────

export const designContractApi = {
  async listPackages(): Promise<DesignSystemDiscovery> {
    return await invoke<DesignSystemDiscovery>("list_design_system_packages_cmd");
  },
  async getPackageContract(packageId: string): Promise<DesignContract> {
    return await invoke<DesignContract>("get_design_system_package_contract_cmd", { packageId });
  },
  async getPackageDetail(packageId: string): Promise<DesignSystemPackageDetail> {
    return await invoke<DesignSystemPackageDetail>("get_design_system_package_detail_cmd", { packageId });
  },
  /** List all built-in design templates (returns [id, name][]). */
  async listTemplates(): Promise<[string, string][]> {
    return await invoke<[string, string][]>("list_design_templates_cmd");
  },

  /** Get a specific built-in template by ID. */
  async getTemplate(templateId: string): Promise<DesignContract> {
    return await invoke<DesignContract>("get_design_template_cmd", {
      templateId,
    });
  },

  /** Compose a design contract from parameters (no disk write). */
  async compose(params: DesignContractParams): Promise<DesignContract> {
    return await invoke<DesignContract>("compose_design_contract_cmd", {
      params,
    });
  },

  /** Preview DESIGN.md output (no disk write). */
  async previewDesignMd(params: DesignContractParams): Promise<string> {
    return await invoke<string>("preview_design_md_cmd", { params });
  },

  /** Preview DTCG JSON output (no disk write). */
  async previewDtcgJson(params: DesignContractParams): Promise<string> {
    return await invoke<string>("preview_dtchg_json_cmd", { params });
  },

  /** Export: compose + write DESIGN.md to project root + archive. */
  async previewExportPlan(
    projectId: string,
    params: DesignContractParams,
  ): Promise<DesignInstallPlan> {
    return await invoke<DesignInstallPlan>("preview_design_export_plan_cmd", {
      projectId,
      params,
    });
  },

  /** Export: compose + write DESIGN.md to project root + archive. */
  async exportContract(
    projectId: string,
    params: DesignContractParams,
  ): Promise<string> {
    return await invoke<string>("export_design_contract_cmd", {
      projectId,
      params,
    });
  },

  /** Preview install plan: pre-flight dry run with audit scan (no disk write). */
  async previewInstallPlan(
    projectId: string,
    params: DesignContractParams,
  ): Promise<DesignInstallPlan> {
    return await invoke<DesignInstallPlan>("preview_design_install_plan_cmd", {
      projectId,
      params,
    });
  },

  /** Install: write DESIGN.md + design-tokens.json (safe, never overwrites). */
  async installContract(
    projectId: string,
    params: DesignContractParams,
  ): Promise<DesignInstallResult> {
    return await invoke<DesignInstallResult>("install_design_contract_cmd", {
      projectId,
      params,
    });
  },

  /** Read-only check that project outputs still match the selected design system. */
  async verifyProjectSystem(projectId: string): Promise<DesignSystemVerification> {
    return await invoke<DesignSystemVerification>("verify_design_system_manifest_cmd", {
      projectId,
    });
  },

  /** Import a DESIGN.md from a local file path. */
  async importFromFile(filePath: string): Promise<ImportResult> {
    return await invoke<ImportResult>("import_design_from_file_cmd", {
      filePath,
    });
  },

  /** Import a DESIGN.md from URL content (content fetched by frontend). */
  async importFromUrl(
    content: string,
    sourceUrl: string,
    sourceKind: "github" | "shadcn",
  ): Promise<ImportResult> {
    return await invoke<ImportResult>("import_design_from_url_cmd", {
      content,
      sourceUrl,
      sourceKind,
    });
  },
};

File: src-tauri/src/commands/design_contract.rs
//! Design Contract commands: template management, compose, export, import, install.

use tauri::State;

use crate::database::Database;
use crate::services::design_contract::{
    compose_design_contract, export_design_contract, generate_design_md, generate_dtchg_json,
    get_design_template, import_design_from_content, import_design_from_file,
    install_design_contract, list_design_templates, preview_export_plan, preview_install_plan,
    verify_design_system_manifest, DesignContract, DesignContractParams, DesignInstallPlan,
    DesignInstallResult, DesignSystemVerification, ImportResult,
};
use crate::services::design_system_registry::{
    discover_design_systems, load_design_system_contract, load_design_system_package_detail,
    DesignSystemDiscovery, DesignSystemPackageDetail,
};
use crate::store::AppState;

fn project_path_for_id(db: &Database, project_id: &str) -> Result<String, String> {
    db.get_project(project_id)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("项目不存在: {project_id}"))
        .map(|p| p.path)
}

/// List all built-in design templates (returns vec of (id, name)).
#[tauri::command]
pub async fn list_design_templates_cmd() -> Result<Vec<(String, String)>, String> {
    Ok(list_design_templates())
}

/// Discover packaged and user-installed offline design-system packages.
#[tauri::command]
pub async fn list_design_system_packages_cmd() -> Result<DesignSystemDiscovery, String> {
    discover_design_systems().map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_design_system_package_contract_cmd(
    package_id: String,
) -> Result<DesignContract, String> {
    load_design_system_contract(&package_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_design_system_package_detail_cmd(
    package_id: String,
) -> Result<DesignSystemPackageDetail, String> {
    load_design_system_package_detail(&package_id).map_err(|e| e.to_string())
}

/// Get a specific built-in template by ID.
#[tauri::command]
pub async fn get_design_template_cmd(template_id: String) -> Result<DesignContract, String> {
    get_design_template(&template_id).ok_or_else(|| format!("模板不存在: {template_id}"))
}

/// Compose a design contract from parameters (no disk write).
#[tauri::command]
pub async fn compose_design_contract_cmd(
    params: DesignContractParams,
) -> Result<DesignContract, String> {
    compose_design_contract(&params).map_err(|e| e.to_string())
}

/// Preview the DESIGN.md output (no disk write).
#[tauri::command]
pub async fn preview_design_md_cmd(params: DesignContractParams) -> Result<String, String> {
    let contract = compose_design_contract(&params).map_err(|e| e.to_string())?;
    generate_design_md(&contract).map_err(|e| e.to_string())
}

/// Preview the DTCG JSON output (no disk write).
#[tauri::command]
pub async fn preview_dtchg_json_cmd(params: DesignContractParams) -> Result<String, String> {
    let contract = compose_design_contract(&params).map_err(|e| e.to_string())?;
    generate_dtchg_json(&contract).map_err(|e| e.to_string())
}

/// Preview overwrite-style export plan (no disk write to project).
#[tauri::command]
pub async fn preview_design_export_plan_cmd(
    state: State<'_, AppState>,
    project_id: String,
    params: DesignContractParams,
) -> Result<DesignInstallPlan, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let contract = compose_design_contract(&params).map_err(|e| e.to_string())?;
    preview_export_plan(&path, &contract).map_err(|e| e.to_string())
}

/// Export: compose + write DESIGN.md to project root + archive in .opensunstar/contract/.
#[tauri::command]
pub async fn export_design_contract_cmd(
    state: State<'_, AppState>,
    project_id: String,
    params: DesignContractParams,
) -> Result<String, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let contract = compose_design_contract(&params).map_err(|e| e.to_string())?;
    export_design_contract(&path, &contract).map_err(|e| e.to_string())
}

/// Preview install plan: pre-flight dry run with audit scan (no disk write to project).
#[tauri::command]
pub async fn preview_design_install_plan_cmd(
    state: State<'_, AppState>,
    project_id: String,
    params: DesignContractParams,
) -> Result<DesignInstallPlan, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let contract = compose_design_contract(&params).map_err(|e| e.to_string())?;
    preview_install_plan(&path, &contract).map_err(|e| e.to_string())
}

/// Install: write DESIGN.md + design-tokens.json to project (safe, never overwrites).
#[tauri::command]
pub async fn install_design_contract_cmd(
    state: State<'_, AppState>,
    project_id: String,
    params: DesignContractParams,
) -> Result<DesignInstallResult, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    let contract = compose_design_contract(&params).map_err(|e| e.to_string())?;
    install_design_contract(&path, &contract).map_err(|e| e.to_string())
}

/// Verify the project files against the selected design-system manifest (read-only).
#[tauri::command]
pub async fn verify_design_system_manifest_cmd(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<DesignSystemVerification, String> {
    let path = project_path_for_id(&state.db, &project_id)?;
    verify_design_system_manifest(&path).map_err(|e| e.to_string())
}

/// Import a DESIGN.md from a local file path.
#[tauri::command]
pub async fn import_design_from_file_cmd(file_path: String) -> Result<ImportResult, String> {
    import_design_from_file(&file_path).map_err(|e| e.to_string())
}

/// Import a DESIGN.md from URL content (content is fetched by frontend, passed here).
#[tauri::command]
pub async fn import_design_from_url_cmd(
    content: String,
    source_url: String,
    source_kind: String,
) -> Result<ImportResult, String> {
    import_design_from_content(&content, &source_url, &source_kind).map_err(|e| e.to_string())
}
